crates/extrinsics/src/contract_storage.rs

```diff
@@ -0,0 +1,310 @@
+                let (label, _) = root_keys.iter().find(|(_, key)| *key == root_key)?;
+
+                Some(ContractStorageCell {
+                    key: k.clone(),
+                    value: v.clone(),
+                    root_key,
+                    mapping_key,
+                    label: label.clone(),
+                })
+            })
+            .collect();
+
+        Self { cells }
+    }
+
+    fn collect_root_keys(
+        label: String,
+        layout: &Layout<PortableForm>,
+        root_keys: &mut Vec<(String, u32)>,
+    ) {
+        match layout {
+            Layout::Root(root) => {
+                root_keys.push((label.clone(), *root.root_key().key()));
+                Self::collect_root_keys(label, root.layout(), root_keys)
+            }
+            Layout::Struct(struct_layout) => {
+                for field in struct_layout.fields() {
+                    let label = field.name().to_string();
+                    Self::collect_root_keys(label, field.layout(), root_keys);
+                }
+            }
+            Layout::Enum(enum_layout) => {
+                for (variant, struct_layout) in enum_layout.variants() {
+                    for field in struct_layout.fields() {
+                        let label =
+                            format!("{}::{}", enum_layout.name(), variant.value());
+                        Self::collect_root_keys(label, field.layout(), root_keys);
+                    }
+                }
+            }
+            Layout::Array(_) => {
+                todo!("Figure out what to do with an array layout")
+            }
+            Layout::Hash(_) => {
+                unimplemented!("Layout::Hash is not currently be constructed")
+            }
+            Layout::Leaf(_) => {}
+        }
+    }
+
+    /// Split the key up
+    ///
+    /// 0x6a3fa479de3b1efe271333d8974501c8e7dc23266dd9bfa5543a94aad824cfb29396d200926d28223c57df8954cf0dc16812ea47
+    /// |--------------------------------|---------|-------------------------------------------------------------|
+    ///       blake2_128 of raw key        root key                         mapping key
+    fn key_parts(key: &Bytes) -> (u32, Option<Bytes>) {
+        assert!(key.0.len() >= 20, "key must be at least 20 bytes");
+        let mut root_key_bytes = [0u8; 4];
+        root_key_bytes.copy_from_slice(&key.0[16..20]);
+
+        // keys are SCALE encoded (little endian), so the root key
+        let root_key = <u32 as scale::Decode>::decode(&mut &root_key_bytes[..])
+            .expect("root key is 4 bytes, it always decodes successfully to a u32; qed");
+
+        let mapping_key = if key.0.len() > 20 {
+            Some(Bytes::from(key.0[20..].to_vec()))
+        } else {
+            None
+        };
+
+        (root_key, mapping_key)
+    }
+}
+
+#[derive(Serialize)]
+pub struct ContractStorageCell {
+    key: Bytes,
+    value: Bytes,
+    label: String,
+    root_key: u32,
```

Comment:
Maybe it would be more useful to dispaly the `root_key` as hex?

Reply:
Possibly, yes - that is how they are serialized in the metadata.

Anyway I will leave that up to you if that is okay? I expect this data structure will not be the final form of how we display the structured storage output.